### ISSUE TITLE
[SuperSocket 1.6.7 and Unity3D] WebSocketServer TLS handshake failure

### ISSUE THREAD

**Issue by ercang** [kerryjiang/SuperSocket#498]
Hi,

first of all, thank you for making this library. I am trying to use it in a project with Unity3D that runs on Ubuntu Linux.

Unity3d does not support .Net Framework 2.1, so I am using SuperSocket version 1.6.7. On my game server, I can use the WebSocket server without any issue if I don't enable TLS. However, if I enable TLS, the handshake does not complete. I tried many things but couldn't find a solution.

- Server: Unity3d, Linux, SuperSocket 1.6.7
- Client: Chrome 90 

This is how I create the server.
```csharp
GameWebSocketServer socketServer = new GameWebSocketServer(); // this is my custom session and appserver
var serverConfig = new SuperSocket.SocketBase.Config.ServerConfig();
serverConfig.Security = SslProtocols.Tls12.ToString();
serverConfig.Certificate = new SuperSocket.SocketBase.Config.CertificateConfig {FilePath = Directory.GetCurrentDirectory() + "/cert.pfx", Password = "", KeyStorageFlags= System.Security.Cryptography.X509Certificates.X509KeyStorageFlags.DefaultKeySet};
serverConfig.MaxConnectionNumber = 100;
serverConfig.Ip = "Any";
serverConfig.Port = port;
socketServer.NewSessionConnected += WebSocketNewSessionConnected;
socketServer.SessionClosed += WebSocketSessionClosed;
socketServer.NewDataReceived += WebSocketNewDataReceived;
socketServer.Setup(serverConfig);
socketServer.Start();
```

I test cert.pfx with another library and it is working. Here are the error logs If I try to connect.
```
2021-05-09 14:39:40,270 [-185661632] DEBUG combatzone.io - Listener (0.0.0.0:8081) was started
2021-05-09 14:39:40,294 [-185661632] INFO  combatzone.io - The server instance combatzone.io has been started!

2021-05-09 14:57:25,022 [Thread Pool Worker] ERROR combatzone.io - Session: 444bec06-56ef-4696-9f19-bf988ed5e7a3/213.227.244.116:49175
Unexpected error
System.Security.Authentication.AuthenticationException: A call to SSPI failed, see inner exception. ---> Mono.Security.Interface.TlsException: Verification failure during handshake
  at Mono.Unity.UnityTlsContext.ProcessHandshake () [0x00081] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
  at Mono.Net.Security.MobileAuthenticatedStream.ProcessHandshake (Mono.Net.Security.AsyncOperationStatus status) [0x0003e] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
  at (wrapper remoting-invoke-with-check) Mono.Net.Security.MobileAuthenticatedStream.ProcessHandshake(Mono.Net.Security.AsyncOperationStatus)
  at Mono.Net.Security.AsyncHandshakeRequest.Run (Mono.Net.Security.AsyncOperationStatus status) [0x00006] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
  at Mono.Net.Security.AsyncProtocolRequest+<ProcessOperation>d__24.MoveNext () [0x000ff] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at Mono.Net.Security.AsyncProtocolRequest+<StartOperation>d__23.MoveNext () [0x0008b] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
   --- End of inner exception stack trace ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at Mono.Net.Security.MobileAuthenticatedStream+<ProcessAuthentication>d__47.MoveNext () [0x00254] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at System.Threading.Tasks.TaskToApm.End (System.IAsyncResult asyncResult) [0x00029] in <44afb4564e9347cf99a1865351ea8f4a>:0 
  at Mono.Net.Security.MobileAuthenticatedStream.EndAuthenticateAsServer (System.IAsyncResult asyncResult) [0x00000] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
  at System.Net.Security.SslStream.EndAuthenticateAsServer (System.IAsyncResult asyncResult) [0x00006] in <6b80e2850f1c48e6a227f0818ca6b227>:0 
  at SuperSocket.SocketEngine.AsyncStreamSocketSession.OnBeginInitStream (System.IAsyncResult result, System.Boolean connect) [0x0002d] in <a0021a983450437388474b6667c46dac>:0 
```

The exception is quite generic and it is generated by this method 
```csharp
AsyncStreamSocketSession::OnBeginInitStream() {
...
sslStream.EndAuthenticateAsServer(result) 
}
```

BeginAuthenticateAsServer() is started with Tls12. It seems that it should be working. I couldn't find where to look for this error.

Here is the test server, https://combatzone.io, and ws uri, wss://combatzone.io:8082/

I appreciate any thoughts on this.

**Comment by chucklu** [comment]
You can try this [How to: Configure network tracing](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/how-to-configure-network-tracing), and review the trace log


If you can not get useful error message from above approach, I would like to  suggest you to learn [the flow of TLS](https://www.acunetix.com/blog/articles/establishing-tls-ssl-connection-part-5/)
And then capture the network packages  by WireShark

**Comment by chucklu** [comment]
@ercang Is it possible that you can provide a sample zip file to test the connection? I can try to analysis the network by WireShark

**Comment by ercang** [comment]
I checked the links you sent. Also investigated it a bit further with Wireshark.

I create the pfx file with the following flags. It is the same certificate that HTTP server uses.
```bash
openssl pkcs12 -inkey privkey.pem -in fullchain.pem -certfile cert.pem -export -out cert.pfx -password pass:
```
I checked with Wireshark, captured it on the server. It seems that the server closes the connection after receives the client key exchange message. It starts with a TCP handshake (syn, syn+ack, ack)

| Src -> Dest  | Info | Detail
| ---------------- | --------------------------- |-----|
|Client -> Server | TLSv1.2 | Client Hello
|Server -> Client | TLSv1.2 | Server Hello | Cipher Suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (0xc030)
|Server -> Client | TLSv1.2 | Certificate, Server Key Exchange, Server Hello Done
|Client -> Server | TLSv1.2 | Client Key Exchange, Change Cipher Spec, Encrypted Handshake Message
|Server -> Client | TCP | FIN, ACK
|Client -> Server | TCP | FIN, ACK

It seems that the handshake stops unexpectedly because the server closes the connection. It should continue like this;

| Src -> Dest  | Info |
| ---------------- | --------------------------- |
| Client -> Server | Client Handshake Finished |
| Server -> Client | Change Cipher Spec            |
| Server -> Client | Server Handshake Finished |

The same process is able to connect to another SSL endpoint successfully. This might have something to do with unity's mono environment but I am not sure.

**Comment by ercang** [comment]
I build the server as a separate C# project with Visual Studio. And it uses the same certificate pfx file and a very similar code, the handshake is completed successfully. The only difference is, I installed mono-develop on Ubuntu server which is a different environment.

There is something different with the environment when it is built with Unity's mono environment. I tested this with Unity 2019.4.19f1, 2019.4.26f1 and 2021.1.6f1. The handshake fails on the server-side.

**Mono** (v 6.12)
- SSL handshake is successful using the same certificate

**Unity**
- It can connect to other wss endpoints as a client
- Fails to complete SSL handshake when it is the server

This is clearly not a problem with SuperSocket, but I will update here if I find a solution :)
